tests/unit/autocomplete/test_completer.py

```diff
@@ -0,0 +1,141 @@
+
+        auto_complete = completer.AutoCompleter(
+            self.parser, completers=[first, second])
+        self.assertEqual(auto_complete.autocomplete('aws e'), ['ec2', 'ecs'])
+
+        first.complete.assert_called_with(self.parsed_result)
+        second.complete.assert_called_with(self.parsed_result)
+
+    def test_returns_empty_list_if_no_completers_have_results(self):
+        first = mock.Mock(spec=completer.BaseCompleter)
+        second = mock.Mock(spec=completer.BaseCompleter)
+
+        first.complete.return_value = None
+        second.complete.return_value = None
+
+        auto_complete = completer.AutoCompleter(
+            self.parser, completers=[first, second])
+        self.assertEqual(auto_complete.autocomplete('aws e'), [])
+
+        first.complete.assert_called_with(self.parsed_result)
+        second.complete.assert_called_with(self.parsed_result)
+
+
+class TestModelIndexCompleter(unittest.TestCase):
+    def setUp(self):
+        self.index = mock.Mock(spec=model.ModelIndex)
+        self.completer = completer.ModelIndexCompleter(self.index)
+
+    def test_does_not_complete_if_unparsed_items(self):
+        parsed = parser.ParsedResult(
+            current_command='aws', lineage=[],
+            unparsed_items=['foo'], last_fragment='')
+        self.assertIsNone(self.completer.complete(parsed))
+
+    def test_does_complete_if_last_fragment_is_none(self):
+        parsed = parser.ParsedResult(
+            lineage=[],
+            unparsed_items=[],
+            current_command='aws',
+            last_fragment=None,
+        )
+        self.assertIsNone(self.completer.complete(parsed))
+
+    def test_can_prefix_match_services(self):
+        parsed = parser.ParsedResult(
+            current_command='aws', lineage=[],
+            last_fragment='e',
+        )
+        self.index.command_names.return_value = ['ec2', 'ecs', 's3']
+        self.assertEqual(self.completer.complete(parsed), ['ec2', 'ecs'])
+
+    def test_returns_all_results_when_last_fragment_empty(self):
+        parsed = parser.ParsedResult(
+            current_command='aws', lineage=[],
+            last_fragment='',
+        )
+        self.index.command_names.return_value = ['ec2', 'ecs', 's3']
+        self.assertEqual(self.completer.complete(parsed), ['ec2', 'ecs', 's3'])
+
+    def test_can_autocomplete_global_param(self):
+        parsed = parser.ParsedResult(
+            lineage=[],
+            unparsed_items=[],
+            current_command=u'aws',
+            last_fragment=u'--re',
+        )
+        self.index.arg_names.return_value = ['region', 'endpoint-url']
+        self.assertEqual(self.completer.complete(parsed), ['--region'])
+
+    def test_can_combine_global_and_command_params(self):
+        parsed = parser.ParsedResult(
+            lineage=['aws', 'ec2'],
+            unparsed_items=[],
+            current_params={},
+            current_command='describe-instances',
+            last_fragment='--r',
```

Comment:
May be a bit off-topic, but do we want to consider autocompleting for options when:

1) The last fragment is an empty string (i.e. not starting with `--`)
2) There are no more commands from the index
3) There are parameters available from the index.

So this would be:
```
aws ec2 describe-instances <tab>
```
and hopefully we would get all of the parameter options out of it?

Reply:
This was an explicit choice to avoid overwhelming the user with too many options.  I think it's pretty intuitive if you want CLI options, you can just type `--`.  There's  a comment that mentions this:

```
            # We could technically offer completion of options
            # if the last fragment is an empty string, but to avoid
            # dumping too much information back to the user, we only
            # offer completions for options if the value starts with
            # '--'.
```

Reply:
I can see that making sense. I would need to actually try a version of the integrated completer to know for sure. I can see it making sense though to do it for required parameters or having a workflow for figuring out the required parameters without having to access the help documentation or just running the command and seeing what validation errors you get. We can explore this more once everything is integrated and we can test it out more.